### ISSUE TITLE
[Java] java-jars fails with "We have duplicated artifacts attached"

### ISSUE THREAD

**Issue by lidavidm** [apache/arrow#41490]
### Describe the bug, including details regarding any error messages, version, and platform.

```
[INFO] <<< source:3.3.0:jar (default-cli) < generate-sources @ module-info-compiler-maven-plugin <<<
[INFO] 
[INFO] 
[INFO] --- source:3.3.0:jar (default-cli) @ module-info-compiler-maven-plugin ---
Error:  We have duplicated artifacts attached.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Arrow Java Root POM 17.0.0-SNAPSHOT:
[INFO] 
[INFO] Arrow Maven Plugins ................................ SUCCESS [ 37.141 s]
[INFO] Module Info Compiler Maven Plugin .................. FAILURE [ 16.713 s]
```

https://github.com/ursacomputing/crossbow/actions/runs/8903205653/job/24451424677#step:6:16247

### Component(s)

Java

**Comment by lidavidm** [comment]
@vibhatha or @jbonofre, any idea? It's unclear when exactly this popped up since the test was blocked by #41470 until just now

**Comment by vibhatha** [comment]
Would it be the recent maven build changes, my best guess would be the last build PR, but I can't be sure. Though I didn't notice it before that.

**Comment by vibhatha** [comment]
Checking my recent crossbow usage 

8 days ago: passes

https://github.com/apache/arrow/pull/40340#issuecomment-2074733574

6 days ago: fails

https://github.com/apache/arrow/pull/40340#issuecomment-2079209380

**Comment by vibhatha** [comment]
I know this doesn't help, but at least it should have happened sometime after 8 days ago.

**Comment by lidavidm** [comment]
Do you think you could bisect it? Or at least figure out which commits are possible causes in that time range?

**Comment by jbonofre** [comment]
That's due to the source maven plugin update coming with new Apache POM. On some profile, the source and jar maven plugins are executed by the parent Apache POM and by the project pom. So we have a dual execution, so dual artifacts. It's what I saw on Apache projects. I gonna take a look if it's the same here in Arrow.

**Comment by vibhatha** [comment]
@lidavidm I am looking into it too. 
But seems like @jbonofre has a good pointer to it.

**Comment by vibhatha** [comment]
@lidavidm I feel like [this](https://github.com/apache/arrow/pull/41211) is the PR which could have broken the CIs. Though java-jars pass in this PR itself, a change made in another PR followed by this merge could have triggered it. It is very hard to predict this since in PR https://github.com/apache/arrow/pull/41309 the java-jars failure is clouded by the C++ filesystem issue which was fixed recently. But that PR has some changes in plugins. But I cannot exactly pin-point what change is causing it. 

cc @jbonofre WDYT?

**Comment by jbonofre** [comment]
@vibhatha I don't think that the plugin-plugin update is the root cause of the problem.
IMHO, I'm suspecting more this commit 9090e679da91e0544171c2da9f2b9ce8ba23d389 as it updated to Apache POM 31.

Let me check locally.

**Comment by vibhatha** [comment]
Yes, this is the PR I am also suspicious about. I am going to wait for your response. Thanks for looking into this @jbonofre

**Comment by jbonofre** [comment]
I think, we hit the same issue as in my PR (https://github.com/apache/arrow/pull/39215). We had similar issue there (due to Apache POM version bump as well). Let me double check if it's the same and eventually leverage the reproducible build PR to fix that 😄

**Comment by lidavidm** [comment]
This will become a release blocker at some point but I suppose 17.0.0 is still a bit off

**Comment by jbonofre** [comment]
@lidavidm agree, so let me tackle that ! At your service 😄

**Comment by lidavidm** [comment]
It's much appreciated :)

**Comment by jbonofre** [comment]
@lidavidm my pleasure, always happy to help 😄

**Comment by jbonofre** [comment]
The problem is in `java_full_build.sh` in crossbow: https://github.com/apache/arrow/blob/c8f89d06bb79909dbd52d7d1504bdcadb220fadd/ci/scripts/java_full_build.sh#L50

It explicitely calls `source:jar` goal whereas it's already done by Apache POM 31.

IMHO, the `java_full_build.sh` should only do:

```
mvn clean \
    install \
    -Papache-release \
    -Parrow-c-data \
    -Parrow-jni \
    -Darrow.cpp.build.dir=$dist_dir \
    -Darrow.c.jni.dist.dir=$dist_dir \
    -DdescriptorId=source-release
```

If agree, I can create a PR to change this pipeline.

**Comment by vibhatha** [comment]
@jbonofre it could be needed in https://github.com/apache/arrow/pull/41772 PR itself?

**Comment by lidavidm** [comment]
Ok, we can test in CI and confirm. Thanks for digging into this!
